src/rest/DataQueryEndpoints.js

```diff
@@ -81,6 +84,9 @@ module.exports = (historicalAdapter, streamFetcher, volumeLogger = new VolumeLog
         const fromTimestamp = parseIntIfExists(req.query.fromTimestamp)
         const toTimestamp = parseIntIfExists(req.query.toTimestamp)
 
+        // TODO: do we just drop offsets (like done below) and keep the rest the same?
```

Comment:
- Let's drop support for `toOffset` and `fromOffset` parameters. We can add a check here which throws if either one is defined: `throw new Error('toOffset and fromOffset parameters are no longer supported. Please use fromTimestamp and toTimestamp')`. Then we can remove all the `toOffset` and `fromOffset` validity checks in the below code.
- Let's try to mirror the `ResendRangeRequest` in this endpoint: `fromTimestamp` and `toTimestamp` must be defined, with `fromSequence`, `toSequence` and `publisherId` optional. If only `fromTimestamp` is defined, you could `throw new Error('fromTimestamp and toTimestamp are required. To request all messages since a timestamp, use the endpoint /streams/:id/data/partitions/:partition/from')`.
- Let's add a new endpoint `/streams/:id/data/partitions/:partition/from` that mirrors the `ResendFromRequest` functionality. Parameter `timestamp` required, with `sequence` and `publisherId` optional.
- Let's update `swagger.json` in `engine-and-editor` to reflect the API changes.

Reply:
What is the idea for having `from` be a separate API endpoint?

Imo it would be more user-friendly to have all query options be part of the same endpoint only differing in query params used.

Reply:
@harbu These are the things I had in mind when I wrote the comment:

- Cleaner code with less `if` branches checking for various valid/invalid combinations of parameters. It can get quite messy with all combinations of `fromTimestamp`, `toTimestamp`, `fromSequence`, `toSequence`, and `publisherId`, most of which are optional, but not valid in arbitrary combinations (`toSequence` doesn't make sense if `toTimestamp` is not defined, etc.)

- Direct analogy to websocket protocol messages

- One endpoint for one use case (`range` for gapfill, `from` for backfill/UI resend)

- Potentially less explaining required for users ("`range` endpoint is for range requests and `from` endpoint is for open-ended requests" vs. explaining that `range` endpoint is for range requests, but it can also do open-ended requests, and in that case you should only define the `from...` fields but none of the `to...` fields)

Reply:
The use case I was thinking about is a user playing around with different values for ` fromTimestamp` and `toTimestamp` and then after gauging the data deciding they only need a lower bound. It will be not enough to remove parameter `toTimestamp` from URL query string but also necessary to change `/range` to `/from` which I would find slightly annoying as a user. Having an error message indicating to the user the need to use `/from` does help though.

Reply:
Yeah, I see! But taking one step back, I guess the user would find out about the endpoints in our API explorer or other docs, meaning that they would probably notice that there are three endpoints `last`, `from`, and `range`, before they start playing around with one of them. The difference between them seems relatively clear from the name.

As for the code, the validation part got cleaner as intended, but to be fair, the change did have the consequence of duplicating some parameter parsing logic which are similar in the endpoints. Works for me either way.